File: IP/AttiCal/AttiCal.py
from z3 import *
import numpy as np
#variables
Gi=[]
Gi_next=[]
W=[]
angle=[]
angle_next=[]
rate=[]
rate_next=[]
VER_Gi=Array('VER_Gi',IntSort(),RealSort())
VER_Gi_next=Array('VER_Gi_next',IntSort(),RealSort())
VER_W=Array('VER_W',IntSort(),RealSort())
VER_W_next=Array('VER_W_next',IntSort(),RealSort())
VER_angle=Array('VER_angle',IntSort(),RealSort())
VER_angle_next=Array('VER_angle_next',IntSort(),RealSort())
VER_rate=Array('VER_rate',IntSort(),RealSort())
VER_rate_next=Array('VER_rate_next',IntSort(),RealSort())
workmode=Int('workmode')
flgSP=Int('flgSP')
piyaw=Int('piyaw')
royaw=Int('royaw')
def Array_length(a):
	temp = np.array(a)
	return temp.size
def List2Array(lst, idx=()):
    if isinstance(lst[0], list):
        size = len(lst)
        z3_array = Array('array_' + '_'.join(map(str, idx)), IntSort(), ArraySort(IntSort(), IntSort()))
        constraints = []
        for i in range(size):
            sub_array, sub_constraints = List2Array(lst[i], idx + (i,))
            constraints.extend(sub_constraints)
            constraints.append(z3_array[i] == sub_array)
        return z3_array, constraints
    else:
        size = len(lst)
        z3_array = Array('array_' + '_'.join(map(str, idx)), IntSort(), IntSort())
        constraints = [z3_array[i] == lst[i] for i in range(size)]
        return z3_array, constraints
s = Solver()
#Contract

File: VeriHar-main/sources/AttiCal.c
//#include "VeriHar-main/include/AttiCal.h"
#include "../include/AttiCal.h"

void AttiCalFun(void *pIp)
{
    AttiCal *p = (AttiCal*)pIp;

	if ( (p->workmode) == SAM_CRUISE )     	/* 巡航方式 */
	{                                                       
		if (p->flgSP == TRUE )   			/* 太阳可见 */
		{                                                   
        	p->angle[0] = p->royaw;	/* 滚动角用太敏 */
        	p->angle[1] = p->piyaw;	/* 俯仰角用太敏 */
        	                                                
        	 												/* 将太敏读数替换当前积分值 */
       		p->Gi[0] = p->royaw;	/* 滚动角陀螺积分  */
       		p->Gi[1] = p->piyaw;	/* 俯仰角陀螺积分  */
		}                                                   
		else												/* 太阳不可见 */
		{                                                   
		    p->angle[0] = p->Gi[0];	/* 滚动角用陀螺积分 */
        	p->angle[1] = p->Gi[1];	/* 俯仰角用陀螺积分 */
		}
	}
	
	else    /* 其他模式 */
	{
		p->angle[0] = 0.0f ;					/* 滚动角清零 */
    	p->angle[1] = 0.0f ;					/* 俯仰角清零 */	
	}                                                       
	                                                        
	p->rate[0] = p->W[0];			/* 滚动角速度用陀螺算出的滚动角速度 */
    p->rate[1] = p->W[1];			/* 俯仰角速度用陀螺算出的俯仰角速度 */
	p->rate[2] = p->W[2];			/* 偏航角速度用陀螺算出的偏航角速度 */
    return;
}
